### PR TITLE
1528 Input File Error Handling

### LINKED ISSUE #1528: Make input file loading less fussy
**Describe the feature / issue**
We currently error and bail out of reading our custom-format input file, which causes no end of drama for users when mistakes happen, or something doesn't get written correctly etc.

We should just carry on reading the input file as best as is possible, and just print error messages along the way.

### PR DESCRIPTION
**Pull request by trisyoungs**
This PR significantly "softens" the assumed severity of errors encountered while reading our (current) input file format. Wherever possible errors are simply reported, but parsing continues, and should greatly reduce the amount of bad UX when loading in a slightly older input file.

Closes #1528.

### CODE REVIEW

**Review by jws-1 — CHANGES_REQUESTED** [review]
I like where this is going, and it works nicely! However, I noticed that when parsing a `Procedure` if a keyword within a Node has an invalid number of parameters, the error propagates through the rest of the sequence. The rest of the sequence will be parsed correctly, but the error messages suggest otherwise.

image

**Comment by trisyoungs** [comment]
Good spot.  Have modified files accordingly to prevent all these spurious errors!

File: src/procedure/nodes/sequence.cpp [hunk 1]
```diff
@@ -511,10 +519,13 @@ bool ProcedureNodeSequence::deserialise(LineParser &parser, const CoreData &core
...
         if (!newNode->deserialise(parser, coreData))
-            return Messenger::error("Failed to read node sequence.\n");
+        {
+            Messenger::error("Failed to read node sequence.\n");
+            errorsEncountered = true;
+        }
     }
 
-    return true;
+    return !errorsEncountered;
```
Comment (jws-1) [inline src/procedure/nodes/sequence.cpp:528]:
```suggestion
    return (!errorsEncountered);
```
For consistency

File: src/procedure/nodes/node.cpp [hunk 2]
```diff
@@ -296,10 +297,10 @@ bool ProcedureNode::deserialise(LineParser &parser, const CoreData &coreData)
         else if (result == KeywordBase::ParseResult::Deprecated)
             Messenger::warn("The '{}' keyword is deprecated and will be removed in a future version.\n", parser.argsv(0));
         else if (result == KeywordBase::ParseResult::Failed)
-            return Messenger::error("Failed to parse keyword '{}'.\n", parser.argsv(0));
+            errorsEncountered = true;
     }
 
-    return true;
+    return !errorsEncountered;
```
Comment (jws-1) [inline src/procedure/nodes/node.cpp:303]:
```suggestion
    return (!errorsEncountered);
```
Consistency

File: src/classes/species_io.cpp [hunk 3]
```diff
@@ -73,10 +73,16 @@ bool Species::read(LineParser &parser, CoreData &coreData)
 
         // Do we recognise this keyword and, if so, do we have an appropriate number of arguments?
         if (!keywords().isValid(parser.argsv(0)))
-            return keywords().errorAndPrintValid(parser.argsv(0));
+        {
+            keywords().errorAndPrintValid(parser.argsv(0));
+            continue;
+        }
```
Comment (jws-1) [inline src/classes/species_io.cpp:79]: Should we also set `errorsEncountered = true` here?

**Review by jws-1 — APPROVED** [review]
Looks good to me now. However, I wonder if we should also set `errorsEncountered` when an invalid keyword is given - for example see my comment on `species_io.cpp`

File: src/classes/species_io.cpp (hunk 3 above)
Comment (trisyoungs) [reply src/classes/species_io.cpp:79]: Yes, and in a bunch of other similar places.

**Pull request merged**